feat: Ensure the article appears exactly as it will when published #1102

dineshsutihar opened the pull request:
Fixes #1102 

## Pull Request details

- Implement exact rendering of articles on the article page, ensuring they appear as the user typed.
- This PR supports backward compatibility with previous articles stored in markdown format, while properly rendering new articles stored in the JSON format.
- Implement DOMPurify to sanitize HTML and protect against XSS attacks.

## Any Breaking changes
- If the "Excerpt" input is left blank in the alpha editor, the default preview will display JSON output. This happens because the article content is stored in JSON format, and without an excerpt, the backend defaults to using the article content for the preview.

## Associated Videos

Preview of new JSON content: 

https://github.com/user-attachments/assets/f9d02c78-08ce-4ff5-b793-68ae69181c73

Preview of old Markdown content:

https://github.com/user-attachments/assets/ae932763-dc4e-41ed-becd-d9570cfed1c4

## Screenshot of issue 
![image](https://github.com/user-attachments/assets/f0711bf8-5f1f-41d6-ac7e-6a80c4d4edb2)

# [Edited] Current Final status: 

https://github.com/user-attachments/assets/815532d9-6dd6-47c2-979f-493fae530599

dineshsutihar committed:
feat: Ensure the article appears exactly as it will when published

dineshsutihar committed:
Update package-lock.json to fix dependency issues

dineshsutihar committed:
Merge branch 'codu-code:develop' into fix-1102-implment-wysiwyg-to-see-article-exactly

dineshsutihar committed:
Merge branch 'develop' into fix-1102-implment-wysiwyg-to-see-article-exactly

dineshsutihar committed:
Refactor JSON parsing to optimize performance by parsing 'post.body' only once

dineshsutihar committed:
fix: add DOMPurify for sanitizing HTML to prevent XSS attacks

NiallJoeMaher commented:
Based on the title "Ensure the article appears exactly as it will when published" I had different expectations. 

The problem is it still looks very different pre publish than in the finished product. 

Something more like this

https://github.com/user-attachments/assets/36b6f305-69d3-450f-9fac-a4bba2f51da1

dineshsutihar commented:
> Based on the title "Ensure the article appears exactly as it will when published" I had different expectations.
> 
> The problem is it still looks very different pre publish than in the finished product.
> 
> Something more like this
> 
>  Screen.Recording.2024-10-19.at.09.03.36.mov

Got it, I'm working on the pre-publish editor design. Could you also review the post-publish video I shared and let me know if there are any issues or improvements needed? @NiallJoeMaher

dineshsutihar committed:
fix: align pre-publish editor design to match post-publish article appearance

dineshsutihar commented:
> Based on the title "Ensure the article appears exactly as it will when published" I had different expectations.
> 
> The problem is it still looks very different pre publish than in the finished product.
> 
> Something more like this
> 
>  Screen.Recording.2024-10-19.at.09.03.36.mov

Hey @NiallJoeMaher , I've updated the pre-publish editor design to exactly match the article view based on your feedback. I've also attached the video showing the final result. Please review and let me know if there are any further issues or improvements you'd like to see.

https://github.com/user-attachments/assets/815532d9-6dd6-47c2-979f-493fae530599

NiallJoeMaher commented:
Hey, thanks for this. But this is a mountain of unneeded code and dependencies. 

We can solve this with just 2 lines of code. 

I was able to fix this one by updating just the 2 parent wrappers of the controller from:

```
<div className="bg-neutral-900 text-white shadow-xl">
    <div className="bg-neutral-900 px-4 py-6 sm:p-6 lg:pb-8">
``` 
To:
```
<div className="mx-auto break-words px-2 pb-4 sm:px-4 md:max-w-3xl">
  <div className="prose mx-auto max-w-3xl dark:prose-invert lg:prose-lg">
    {!body && (
      <Controller
        name="body"
        control={control}
        render={({ field }) => (
          <Editor {...field} initialValue={"{}"} />
        )}
      />
    )}
...
```

dineshsutihar commented:
These two lines work perfectly in dark mode, but when you switch to light mode, you might notice that a `div` with `className="bg-black"` is causing issues, as shown below:

![image](https://github.com/user-attachments/assets/160c1837-be00-475c-8d13-7b6f77727312)

### What I’ve done for the editor page (pre-publish page):
- I added the same two lines as yours.
- I removed an unnecessary `div` with `className="bg-black"`, which was causing the issue. Due to this removal, the code inside was shifted one tab to the left, making it look like more changes than there actually are. In reality, it's just the addition of two lines and the removal of one.
- Removed unnecessary imports to streamline the code (1 removed).


### Changes made for the article page (post-publish page) and details on dependencies:

- **Add two packages: `@tiptap/html` and `isomorphic-dompurify`**
    - **`@tiptap/html`**: [Recommended by Tiptap](https://tiptap.dev/docs/guides/output-json-html#render) for parsing the JSON content (generated by the editor) into HTML on the server-side. 
    - **`isomorphic-dompurify`**: Used to sanitize the parsed HTML before injecting it into the DOM. This is crucial to mitigate the risk of XSS (Cross-Site Scripting) attacks, ensuring that no malicious scripts are executed when displaying user-generated content.

- **Modified `components/editor/editor/extensions/index.tsx`**:
    - Removes the `SlashCommand` extension from this file because it was not being configured and it was also a client-side component that caused issues when trying to use it in a server-side context. Now we can use Extensions anywere in application. 

- **Changes in `app/(app)/articles/[slug]/page.tsx`**:
    - Database contains articles in both Markdown (old data) and JSON (current data), I updated the code to handle both formats. It now checks if the data is valid JSON; if so, it uses the new parsing method, otherwise, it falls back to the old Markdown parsing using `@markdoc/markdoc`.
    - Sanitized the parsed HTML with `isomorphic-dompurify`.

- **Reason for changes**:
    - There was no method for parsing JSON, which caused the articles to display incorrectly, as shown below:  
    ![image](https://github.com/user-attachments/assets/572f3394-d412-40d4-9dc5-fe66e3ae6565)


Is there anything else you'd like me to adjust, or does this solution work for you? Let me know if you have any questions or concerns! @NiallJoeMaher

dineshsutihar commented:
These two lines work perfectly in dark mode, but when you switch to light mode, you might notice that a `div` with `className="bg-black"` is causing issues, as shown below :

![image](https://github.com/user-attachments/assets/160c1837-be00-475c-8d13-7b6f77727312)

### What I’ve done for the editor page (pre-publish page):
- I added the same two lines as yours.
I removed an unnecessary `div` with `className="bg-black"`, which was causing the issue. Due to this removal, the code inside was shifted one tab to the left, making it look like there were more changes than there actually were. In reality, it's just the addition of two lines and the removal of one. 
- Removed unnecessary imports to streamline the code (1 removed).


### Changes made for the article page (post-publish page) and details on dependencies:

- **Add two packages: `@tiptap/html` and `isomorphic-dompurify`**
    - **`@tiptap/html`**: [Recommended by Tiptap](https://tiptap.dev/docs/guides/output-json-html#render) for parsing the JSON content (generated by the editor) into HTML on the server-side. 
    - **`isomorphic-dompurify`**: Used to sanitize the parsed HTML before injecting it into the DOM. This is crucial to mitigate the risk of XSS (Cross-Site Scripting) attacks, ensuring that no malicious scripts are executed when displaying user-generated content.

- **Modified `components/editor/editor/extensions/index.tsx`**:
    - Removes the `SlashCommand` extension from this file because it was not being configured and it was also a client-side component that caused issues when trying to use it in a server-side context. Now we can use Extensions anywere in application. 

- **Changes in `app/(app)/articles/[slug]/page.tsx`**:
    - Database contains articles in both Markdown (old data) and JSON (current data), I updated the code to handle both formats. It now checks if the data is valid JSON; if so, it uses the new parsing method, otherwise, it falls back to the old Markdown parsing using `@markdoc/markdoc`.
    - Sanitized the parsed HTML with `isomorphic-dompurify`.

- **Reason for changes**:
    - There was no method for parsing JSON, which caused the articles to display incorrectly, as shown below:  
    ![image](https://github.com/user-attachments/assets/572f3394-d412-40d4-9dc5-fe66e3ae6565)


Is there anything else you'd like me to adjust, or does this solution work for you? Let me know if you have any questions or concerns! @NiallJoeMaher

dineshsutihar commented:
The following five `div`s elements can also be removed:

```jsx
<div className="mx-auto w-full max-w-7xl flex-grow text-black lg:flex xl:px-8">
  {/* Left sidebar & main wrapper */}
  <div className="min-w-0 flex-1 xl:flex">
    <div className="lg:min-w-0 lg:flex-1">
      <div className="h-full min-h-[40rem] px-4 py-0 sm:px-6 lg:px-4">
        {/* Start main area */}
        <div className="relative h-full">
          {/* Main editor code that is unaffected by the above changes */}
        </div>
        {/* End main area */}
      </div>
    </div>
  </div>
</div>
```
What do you think of this? Should I remove it as well?

NiallJoeMaher commented:
@dineshsutihar, thanks for this! Absolutely fantastic notes, and a lot of it makes sense now that I understand it. 🦾

The simpler fix (using the `.prose` class) for the moment will ensure the two stay in sync. Old and new editors (thanks for pointing out the light/issue.)

We have an issue open to fix the slash command (#1130). I think it was accidentally removed in a recent PR. It's an important feature, so keeping it for now will make that issue easier (or one you can tackle after this). Or if this one fixes this issue link it in the PR too! (You can split it if you want 2 Hacktoberfest points).

I like how you manage the two different previews (for markdown or JSON). And makes sense on the two packages you have added for this now that you have explained it. Cheers 🍻 

I see the excerpt issue. It's not a problem for the minute because we haven't updated the backend. The plan for saving the data is to save the raw text (for search) without JSON. The excerpt should be raw text (no JSON or markdown). We can make an issue if you want to continue work on it. 

In the future, we will have a markdown column, a JSON column, and a raw text column, which should make it easier for the logic (for choosing which parsing tool). I will write a script to migrate all old content to JSON so it's consistent too. 

If this PR tackles some side issues, let's open up some separate issues, and I can assign them to you, too (the smaller the PR, the easier it is to review).

NiallJoeMaher commented:
> The following five `div`s elements can also be removed:
> 
> 
> 
> ```jsx
> 
> <div className="mx-auto w-full max-w-7xl flex-grow text-black lg:flex xl:px-8">
> 
>   {/* Left sidebar & main wrapper */}
> 
>   <div className="min-w-0 flex-1 xl:flex">
> 
>     <div className="lg:min-w-0 lg:flex-1">
> 
>       <div className="h-full min-h-[40rem] px-4 py-0 sm:px-6 lg:px-4">
> 
>         {/* Start main area */}
> 
>         <div className="relative h-full">
> 
>           {/* Main editor code that is unaffected by the above changes */}
> 
>         </div>
> 
>         {/* End main area */}
> 
>       </div>
> 
>     </div>
> 
>   </div>
> 
> </div>
> 
> ```
> 
> What do you think of this? Should I remove it as well?

Yes if we can without it causing any problems. Thanks! Tag me when it's all re-ready to review.

dineshsutihar committed:
Merge branch 'develop' into fix-1102-implment-wysiwyg-to-see-article-exactly

dineshsutihar committed:
refactor: remove unnecessary divs and clean up code without affecting layout

dineshsutihar commented:
Hey @NiallJoeMaher, I've refactored the code without affecting the current layout and ensured the pre-publish editor stays in sync with the post-publish article. Can you review it?

- The preview handling (Markdown vs JSON) is working as expected.
- I'll tackle the slash command issue in a separate PR.
- The editor now uses the same code as the snippet below, with unnecessary code removed without affecting its layout and functionallity :
```jsx
<div className="bg-neutral-900 text-white shadow-xl">
  <div className="bg-neutral-900 px-4 py-6 sm:p-6 lg:pb-8">
```

## Here’s the screenshot of the output:  
### Editor Page (Pre-publish):
![image](https://github.com/user-attachments/assets/58d82308-6108-49b3-a206-0d5a90c978b7)

### Article Page (Post-publish):
![image](https://github.com/user-attachments/assets/c4ae9948-1143-4aca-95be-ed227c68174f)

NiallJoeMaher committed:
Merge branch 'develop' into fix-1102-implment-wysiwyg-to-see-article-exactly

NiallJoeMaher commented:
Eslint and build failure but looks good from a glance. Once the checks pass I'll give a full review.

dineshsutihar committed:
fix: lint issue by replacing the type form any to JSONContent

dineshsutihar committed:
Merge branch 'develop' into fix-1102-implment-wysiwyg-to-see-article-exactly

dineshsutihar commented:
Fixed all ESLint and Prettier formatting errors. 

@NiallJoeMaher Can you give full review?

dineshsutihar commented on app/(app)/articles/[slug]/page.tsx:
```diff
@@ -66,11 +85,25 @@ const ArticlePage = async ({ params }: Props) => {
+    return notFound();
+  }
+
+  const parsedBody = parseJSON(post.body);
+  const isTiptapContent = parsedBody?.type === "doc";
+
+  let renderedContent: string | RenderableTreeNode;
+
+  if (isTiptapContent && parsedBody) {
+    const jsonContent = parsedBody;
+    renderedContent = renderSanitizedTiptapContent(jsonContent);
+  } else {
+    const ast = Markdoc.parse(post.body);
+    const transformedContent = Markdoc.transform(ast, config);
+    renderedContent = Markdoc.renderers.react(transformedContent, React, {
+      components: markdocComponents,
+    }) as unknown as string;
```
Keep the type assertion because `renderedContent` needs to be treated as a string for use in `dangerouslySetInnerHTML`, and removing it could lead to runtime issues.

NiallJoeMaher committed:
Merge branch 'develop' into fix-1102-implment-wysiwyg-to-see-article-exactly

NiallJoeMaher reviewed (dismissed):
Great work! Excited to get this merged. 🌮

NiallJoeMaher commented:
Just getting prettier issues by the looks of things. Hoping the build will pass 🤞

dineshsutihar commented:
I ran `ESLint` and `Prettier` commands multiple times before pushing the code and didn't encounter any Prettier errors. Let me install Prettier globally to see if that fixes the issue.

dineshsutihar committed:
style: fix prettier formatting issues

dineshsutihar commented:
I've fixed the Prettier issues, and the code should be good now. Hopefully, all the tests will pass this time! 🤞 @NiallJoeMaher

NiallJoeMaher reviewed (approved):
🌮 🦖  Nice one!
